sidechain/contracts/group/Group.sol

```diff
@@ -309,31 +448,210 @@ contract Group is GroupI, UsingExternalStorage
         pure
         returns (bytes32)
     {
-        return keccak256(abi.encodePacked(_groupId, _role, _secretHash, _nonce));
+        return keccak256(
+            abi.encodePacked(
+                _groupId, 
+                "storeInvitation", // for sig uniqueness
+                _role, 
+                _secretHash, 
+                _nonce
+            )
+        );
     }
 
-    function inviteToCommunity(
+    // store invitation 
+    function storeInvitation(
         uint256 _groupId, 
         uint8   _role,
         bytes32 _secretHash,
         bytes memory _signature,
         uint256 _nonce
     )
         public
-        returns (bool) // the doc says return txHash, so we'll wait for tx receipt?
+        returns (bool) 
     {
-        address _signer = getSigner(_secretHash, _signature, _nonce);
-        addMember(_groupId, _signer, _role);
+        // recover signer, and set as address
+        storageContract.putAddressValue(keccak256(
+            abi.encodePacked(INVITATION_KEY, _groupId, _secretHash, "SIGNER")), 
+            getSigner(_secretHash, _signature, _nonce)
```

Comment:
This cannot work because that line tries to recover the signer from the signature and the message.
```
signature=ecsign(private_key_signer, message)
public_key_signer=ecrecover(signature, message)
```

But the message signed beforehand isn't `_secretHash` but `keccak256(abi.encodePacked(_groupId,  "storeInvitation",_role, _secretHash, _nonce))` from `prepareInvitation`; so ecrecover will recover another random address which likely has a nonce=0.

You need to internally call `prepareInvitation` to get the same message as signed and recover the signature to the right message
``` 
// recover signer, and set as address
storageContract.putAddressValue(keccak256(
    abi.encodePacked(INVITATION_KEY, _groupId, _secretHash, "SIGNER")), 
    getSigner(prepareInvitation(_groupId,  _role, _secretHash, _nonce)), _signature, _nonce);
```

----------------------------------------

Do not hesitate to separate the interfaces/public functions from the logic for more clarity:

```
// Meta-tx interface
function storeInvitation(uint256 _groupId, uint8 _role, bytes32 _secretHash, bytes memory _signature, uint256 _nonce) public return (bool){
    address signer = getSigner(prepareInvitation(_groupId, _role, _secretHash, _nonce)), _signature, _nonce);
    storeInvitation(_groupId, _role, _secretHash, signer); 
}

// Common interface
function storeInvitation(uint256 _groupId, uint8 _role, bytes32 _secretHash) public return (bool) {
    storeInvitation(_groupId, _role, _secretHash, msg.sender); 
}

// Logic implementation
function storeInvitation(uint256 _groupId, uint8   _role, bytes32 _secretHash, address _sender) internal return (bool) {
    (...)
}
```

Reply:
discussed on call, will start working on restructuring contracts to GroupI / GroupConnector / GroupLogic late next week